### PR TITLE
Refactor call & function handling in trans, enable MIR bootstrap.

### LINKED ISSUE #10116: remove the need for foreign ABI wrappers
To generate code for `extern "C" fn foo()`, `rustc` will create a Rust function and wrap it inside a function exposing the C ABI. This is bad for compile-time, but *more importantly* it makes it very hard to write functions like `memset`, `memcpy` and `memmove` for freestanding Rust because LLVM will kindly replace their body with a call to the intrinsic, and then lower that to an infinitely recursive function.

https://github.com/thestinger/rust-core#freestanding

### LINKED ISSUE #30235: Feature-gate non-Rust ABI methods
Currently we allow methods to be marked with the `extern` qualifier indicating that they have a non-Rust ABI (default is "C"). They are rather bug-ridden, only really working by accident, see #26997 and #26049.

While I personally think they shouldn't be allowed to start with, putting them behind a feature-gate until we make a decision either way seems like a good idea.

Hopefully this doesn't require an RFC, but if it does, I'll write one up and this can be a tracking issue.

/cc @rust-lang/lang

### PR DESCRIPTION
**Pull request by eddyb**
Non-Rust and Rust ABIs were combined into a common codepath, which means:
- The ugly `__rust_abi` "clown shoes" shim for C->Rust FFI is gone, fixes #10116.
- Methods, _including virtual ones_ support non-Rust ABIs, closes #30235.
- Non-Rust ABIs also pass fat pointers in two arguments; the result should be identical.
- Zero-sized types are never passed as arguments; again, behavior shouldn't change.

Additionally, MIR support for calling intrinsics (through old trans) was implemented.
Alongside assorted fixes, it enabled MIR to launch :rocket: and do a _complete_ bootstrap.
To try it yourself, `./configure --enable-orbit` _or_ `make RUSTFLAGS="-Z orbit"`.

### CODE REVIEW

**Comment by alexcrichton** [comment]
r? @nikomatsakis 

(just making sure an assignment is given, feel free to delegate elsewhere)

**Comment by eddyb** [comment]
Status: I can build `snap-stage3` with MIR forced on *for everything*.
These are the common run-pass failures for stage1 and stage2:
(that means they are likely MIR trans bugs and not due to miscompiled rustc)
```
failures:
    [run-pass] run-pass/allocator-override.rs
    [run-pass] run-pass/asm-in-out-operand.rs
    [run-pass] run-pass/asm-indirect-memory.rs
    [run-pass] run-pass/asm-out-assign.rs
    [run-pass] run-pass/associated-const-type-parameters.rs
    [run-pass] run-pass/backtrace-debuginfo.rs
    [run-pass] run-pass/const-str-ptr.rs
    [run-pass] run-pass/ignore-all-the-things.rs
    [run-pass] run-pass/int-abs-overflow.rs
    [run-pass] run-pass/issue-14936.rs
    [run-pass] run-pass/issue-15080.rs
    [run-pass] run-pass/issue-15104.rs
    [run-pass] run-pass/issue-23338-ensure-param-drop-order.rs
    [run-pass] run-pass/issue-7784.rs
    [run-pass] run-pass/issue-8460.rs
    [run-pass] run-pass/issue-8860.rs
    [run-pass] run-pass/match-vec-alternatives.rs
    [run-pass] run-pass/mir_raw_fat_ptr.rs
    [run-pass] run-pass/num-wrapping.rs
    [run-pass] run-pass/out-of-stack.rs
    [run-pass] run-pass/simd-intrinsic-generic-cast.rs
    [run-pass] run-pass/simd-intrinsic-generic-elements.rs
    [run-pass] run-pass/super-fast-paren-parsing.rs
    [run-pass] run-pass/vec-matching-fixed.rs
    [run-pass] run-pass/vec-matching-fold.rs
    [run-pass] run-pass/vec-matching-legal-tail-element-borrow.rs
    [run-pass] run-pass/vec-matching.rs
    [run-pass] run-pass/vec-tail-matching.rs
    [run-pass] run-pass/zero-size-type-destructors.rs
```
The *only* run-pass test that only fails in stage2 is `run-pass/super-fast-paren-parsing.rs`, which ends up in a stack overflow in `rustc` (@arielb1 was talking about needing llvm lifetime intrinsics to prevent this, I think).

@rust-lang/tools If I provide a flag to enable MIR trans for everything, and an attribute to mark functions which can't be yet properly translated via MIR, would it be possible to have buildbot ensure MIR doesn't regress?

**Comment by nikomatsakis** [comment]
:bangbang: :astonished:

**Comment by alexcrichton** [comment]
@eddyb I'd be fine at least adding a MIR bot! If it's a `./configure` switch just lemme know and I can set up something like `auto-linux-64-opt-mir` or something like that.

Do you think MIR will want platform coverage as well? A Linux/OSX/Windows builder, or just one with it enabled?

**Comment by eddyb** [comment]
@alexcrichton Before the next snapshot, `make RUSTFLAGS_STAGE1="-Z orbit" RUSTFLAGS_STAGE2=-"Z orbit"`, after that just `make RUSTFLAGS="-Z orbit"`.

Sadly, it doesn't look like tests inherit `RUSTFLAGS_STAGEn`, just `RUSTFLAGS`, so we might need a snapshot *anyway*.

We'll want at least a non-MSVC and a MSVC builder, because of the differences in how unwinding works there.

**Comment by alexcrichton** [comment]
Ok, sounds like a linux + MSVC builder, and can you add this as a `./configure` option? That'll be easiest to add to our build matrix.

**Comment by eddyb** [comment]
@alexcrichton Does that last commit look good?

**Comment by alexcrichton** [comment]
Looks good to me!

**Comment by eddyb** [comment]
Down to 9 run-pass MIR failures:
```
    [run-pass] run-pass/backtrace-debuginfo.rs
    [run-pass] run-pass/const-str-ptr.rs
    [run-pass] run-pass/int-abs-overflow.rs
    [run-pass] run-pass/issue-23338-ensure-param-drop-order.rs
    [run-pass] run-pass/issue-8460.rs
    [run-pass] run-pass/issue-8860.rs
    [run-pass] run-pass/mir_raw_fat_ptr.rs
    [run-pass] run-pass/simd-intrinsic-generic-elements.rs
    [run-pass] run-pass/zero-size-type-destructors.rs
```

I'll just start tagging them with `#[rustc_no_mir]`, this PR is getting kind of long...

**Comment by eddyb** [comment]
@alexcrichton Something changed on master and now, while it still bootstraps `-Z orbit`, there are issues with testing - I have to manually add `--sysroot=x86_64-unknown-linux-gnu/stage2` to `libtest`, `compiletest`, `librustdoc` and `rustdoc` that `stage1` produces, just to run `make check-stage2`, otherwise I get hundreds of these:
```
error: cannot satisfy dependencies so `std` only shows up once
help: having upstream crates all available in one format will likely make this go away
```

And `check-stage2-run-pass-full` fails, with linker errors (I'm expecting a similar cause - either that or it's the NixOS I'm running it on).

`rustdoc` tests, of all things, fail with:
```
: CommandLine Error: Option 'interpreter-print-volatile' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options
```

The only other test failures (with `-Z orbit`) are some `run-make` bugs, which I would look into, but at this point it's too annoying to run everything manually and I don't know how to debug the "incompatible crates" issue (it *could* maybe be MIR mistreating some `enum` in rustc, but which one?).

**Comment by eddyb** [comment]
@alexcrichton @nikomatsakis To be clear, everything should still work without `RUSTFLAGS="-Z orbit"`.

**Comment by eddyb** [comment]
@alexcrichton Looks like parallelism shennanigans: `rm x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/stamp.{getopts,log,term}` lets me compile `libtest` and `compiletest` - however, I did have to trigger a `liblog` build manually, since `compiletest` doesn't appear to have an explicit dependency on it.

**Comment by eddyb** [comment]
@alexcrichton Sorry for all that noise, looks like parallelism got the better of me.
I'll have to wait for *yet another* rebuild, but AFAICT `make check-stage2` completes with `-Z orbit`.

**Comment by eddyb** [comment]
I just did a quick check of MIR trans vs old trans. MIR trans is faster by half a second (or 3.5%) but there's some memory usage discrepancy that I don't know about - if it were extra LLVM IR, it would get optimized away, presumably, but it appears to stick "on top" of the LLVM IR memory usage and doesn't go away when the Rust contexts are dropped, either:
```
eddy@build ~/rust> make x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/stamp.rustc RUSTFLAGS="-Z time-passes"
cfg: version 1.9.0-dev (a67b28097 2016-03-12)
cfg: build triple x86_64-unknown-linux-gnu
cfg: host triples x86_64-unknown-linux-gnu
cfg: target triples x86_64-unknown-linux-gnu
cfg: host for x86_64-unknown-linux-gnu is x86_64
cfg: os for x86_64-unknown-linux-gnu is unknown-linux-gnu
cfg: have good valgrind for x86_64-unknown-linux-gnu
cfg: disabling valgrind run-pass tests
rustc: x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc
time: 0.236; rss: 78MB  parsing
time: 0.084; rss: 78MB  configuration 1
time: 0.000; rss: 78MB  recursion limit
time: 0.006; rss: 78MB  gated macro checking
time: 0.000; rss: 78MB  crate injection
time: 0.015; rss: 82MB  macro loading
time: 0.000; rss: 82MB  plugin loading
time: 0.000; rss: 82MB  plugin registration
time: 1.096; rss: 120MB expansion
time: 0.037; rss: 120MB complete gated feature checking 1
time: 0.223; rss: 120MB configuration 2
time: 0.000; rss: 120MB gated configuration checking
time: 0.105; rss: 120MB maybe building test harness
time: 0.099; rss: 121MB prelude injection
time: 0.018; rss: 121MB checking that all macro invocations are gone
time: 0.000; rss: 121MB checking for inline asm in case the target doesn't support it
time: 0.037; rss: 121MB complete gated feature checking 2
time: 0.018; rss: 121MB const fn bodies and arguments
time: 0.099; rss: 121MB assigning node ids
time: 0.164; rss: 189MB lowering ast -> hir
time: 0.052; rss: 199MB indexing hir
time: 0.019; rss: 199MB attribute checking
time: 0.070; rss: 199MB early lint checks
time: 0.014; rss: 232MB external crate/lib resolution
time: 0.016; rss: 236MB language item collection
time: 0.242; rss: 285MB resolution
time: 0.024; rss: 285MB lifetime resolution
time: 0.000; rss: 285MB looking for entry point
time: 0.000; rss: 285MB looking for plugin registrar
time: 0.084; rss: 304MB region resolution
time: 0.016; rss: 304MB loop checking
time: 0.016; rss: 304MB static item recursion checking
time: 0.087; rss: 311MB type collecting
time: 0.003; rss: 311MB variance inference
time: 0.082; rss: 328MB coherence checking
time: 0.490; rss: 333MB wf checking
time: 0.288; rss: 335MB item-types checking
time: 9.299; rss: 465MB item-bodies checking
time: 0.000; rss: 465MB drop-impl checking
time: 0.591; rss: 467MB const checking
time: 0.181; rss: 467MB privacy checking
time: 0.022; rss: 467MB stability index
time: 0.049; rss: 467MB intrinsic checking
time: 0.032; rss: 467MB effect checking
time: 0.157; rss: 467MB match checking
time: 0.087; rss: 467MB liveness checking
time: 0.622; rss: 467MB rvalue checking
time: 0.940; rss: 498MB MIR dump
time: 0.000; rss: 498MB MIR passes
time: 1.090; rss: 500MB borrow checking
time: 0.020; rss: 501MB reachability checking
time: 0.089; rss: 501MB death checking
time: 0.080; rss: 503MB stability checking
time: 0.000; rss: 503MB unused lib feature checking
time: 0.473; rss: 503MB lint checking
time: 0.002; rss: 503MB resolving dependency formats
time: 0.098; rss: 509MB erasing regions from MIR
  time: 2.262; rss: 550MB       translation item collection
  time: 1.118; rss: 1152MB      write metadata
time: 14.097; rss: 1152MB       translation
  time: 5.218; rss: 882MB       llvm function passes [0]
  time: 96.808; rss: 853MB      llvm module passes [0]
          time: 24.214; rss: 857MB      codegen passes [0]
  time: 0.004; rss: 857MB       codegen passes [0]
time: 128.125; rss: 850MB       LLVM passes
  time: 0.343; rss: 850MB       running linker
time: 1.060; rss: 850MB linking
eddy@build ~/rust> rm x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/stamp.rustc
eddy@build ~/rust> make x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/stamp.rustc RUSTFLAGS="-Z time-passes -Z orbit"
cfg: version 1.9.0-dev (a67b28097 2016-03-12)
cfg: build triple x86_64-unknown-linux-gnu
cfg: host triples x86_64-unknown-linux-gnu
cfg: target triples x86_64-unknown-linux-gnu
cfg: host for x86_64-unknown-linux-gnu is x86_64
cfg: os for x86_64-unknown-linux-gnu is unknown-linux-gnu
cfg: have good valgrind for x86_64-unknown-linux-gnu
cfg: disabling valgrind run-pass tests
rustc: x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc
time: 0.233; rss: 77MB  parsing
time: 0.084; rss: 77MB  configuration 1
time: 0.000; rss: 77MB  recursion limit
time: 0.006; rss: 77MB  gated macro checking
time: 0.000; rss: 77MB  crate injection
time: 0.014; rss: 82MB  macro loading
time: 0.000; rss: 82MB  plugin loading
time: 0.000; rss: 82MB  plugin registration
time: 1.075; rss: 119MB expansion
time: 0.038; rss: 119MB complete gated feature checking 1
time: 0.226; rss: 119MB configuration 2
time: 0.000; rss: 119MB gated configuration checking
time: 0.106; rss: 119MB maybe building test harness
time: 0.102; rss: 120MB prelude injection
time: 0.019; rss: 120MB checking that all macro invocations are gone
time: 0.000; rss: 120MB checking for inline asm in case the target doesn't support it
time: 0.038; rss: 120MB complete gated feature checking 2
time: 0.018; rss: 120MB const fn bodies and arguments
time: 0.100; rss: 120MB assigning node ids
time: 0.160; rss: 188MB lowering ast -> hir
time: 0.052; rss: 198MB indexing hir
time: 0.019; rss: 198MB attribute checking
time: 0.071; rss: 198MB early lint checks
time: 0.013; rss: 231MB external crate/lib resolution
time: 0.016; rss: 235MB language item collection
time: 0.239; rss: 284MB resolution
time: 0.025; rss: 284MB lifetime resolution
time: 0.000; rss: 284MB looking for entry point
time: 0.000; rss: 284MB looking for plugin registrar
time: 0.082; rss: 302MB region resolution
time: 0.016; rss: 302MB loop checking
time: 0.016; rss: 302MB static item recursion checking
time: 0.086; rss: 310MB type collecting
time: 0.002; rss: 310MB variance inference
time: 0.083; rss: 325MB coherence checking
time: 0.487; rss: 332MB wf checking
time: 0.277; rss: 334MB item-types checking
time: 9.033; rss: 463MB item-bodies checking
time: 0.000; rss: 463MB drop-impl checking
time: 0.581; rss: 464MB const checking
time: 0.181; rss: 464MB privacy checking
time: 0.021; rss: 464MB stability index
time: 0.048; rss: 464MB intrinsic checking
time: 0.032; rss: 464MB effect checking
time: 0.153; rss: 464MB match checking
time: 0.085; rss: 465MB liveness checking
time: 0.613; rss: 465MB rvalue checking
time: 0.930; rss: 497MB MIR dump
time: 0.000; rss: 497MB MIR passes
time: 1.081; rss: 499MB borrow checking
time: 0.020; rss: 500MB reachability checking
time: 0.088; rss: 500MB death checking
time: 0.080; rss: 502MB stability checking
time: 0.000; rss: 502MB unused lib feature checking
time: 0.477; rss: 502MB lint checking
time: 0.002; rss: 502MB resolving dependency formats
time: 0.098; rss: 508MB erasing regions from MIR
  time: 2.258; rss: 549MB       translation item collection
  time: 1.123; rss: 1311MB      write metadata
time: 13.410; rss: 1311MB       translation
  time: 6.678; rss: 1045MB      llvm function passes [0]
  time: 123.685; rss: 1050MB    llvm module passes [0]
  time: 31.072; rss: 1054MB     codegen passes [0]
  time: 0.004; rss: 987MB       codegen passes [0]
time: 163.689; rss: 981MB       LLVM passes
  time: 0.350; rss: 981MB       running linker
time: 1.169; rss: 985MB linking
```

**Comment by Aatch** [comment]
@eddyb looking at that breakdown, the extra memory is related to metadata, is there extra metadata being written?

**Comment by eddyb** [comment]
@Aatch I guess I should compare the sizes of the metadata files in the `rlib`s, right?

**Comment by eddyb** [comment]
Old trans:
```
rw-r--r-- 0/0 7295904 Jan  1 00:00 1970 rustc-9026086f.0.o
rw-r--r-- 0/0 15070307 Jan  1 00:00 1970 rust.metadata.bin
rw-r--r-- 0/0 11801988 Jan  1 00:00 1970 rustc-9026086f.0.bytecode.deflate
```
MIR trans (`-Z orbit`):
```
rw-r--r-- 0/0 9368656 Jan  1 00:00 1970 rustc-9026086f.0.o
rw-r--r-- 0/0 15067387 Jan  1 00:00 1970 rust.metadata.bin
rw-r--r-- 0/0 13306872 Jan  1 00:00 1970 rustc-9026086f.0.bytecode.deflate
```
Metadata seems to have gotten *slightly* smaller (weird because it should be identical), but the IR and object file are indeed larger.

**Comment by nikomatsakis** [comment]
r=me modulo the occasional nit. This is great stuff. Everything seems pretty clean. Unless I missed something, the thrust of this is:

1. Centralizing all the code around calls into `Callee` and `FnType`, enabling us to get rid of a bunch of shims
2. A variety of small fixes to MIR and other misc cleanup
3. Some switches and things

**Comment by nikomatsakis** [comment]
@eddyb 

> The correct solution is to have trans::collector produce a list of items to be translated - including closures, instead of recursing the HIR for items and calling trans_fn.

Yes. This has been @michaelwoerister's plan for some time now.

**Comment by nikomatsakis** [comment]
@eddyb 

> Would putting it in trans/mir/mod.rs work for you?

Sure... I guess if it's just a type def'n it's fine, though collector does seem sort of wrong.

File: src/librustc_trans/trans/collector.rs [hunk 1]
```diff
@@ -434,20 +418,18 @@ fn check_recursion_limit<'a, 'tcx: 'a>(ccx: &CrateContext<'a, 'tcx>,
     // more than the recursion limit is assumed to be causing an
     // infinite expansion.
     if recursion_depth > ccx.sess().recursion_limit.get() {
-        if let Some(node_id) = ccx.tcx().map.as_local_node_id(def_id) {
-            ccx.sess().span_fatal(ccx.tcx().map.span(node_id),
-                "reached the recursion limit during monomorphization");
+        let error = format!("reached the recursion limit while instantiating {}",
```
Comment (nikomatsakis) [inline src/librustc_trans/trans/collector.rs]: Nit: while instantiating `{}`

File: src/librustc_trans/trans/abi.rs [hunk 2]
```diff
@@ -67,6 +67,18 @@ pub struct ArgType {
...
+    /// `struct Foo(fn(Foo));`, `original_ty` is still LLVM's `%Foo = {}`.
+    /// The field type will likely end up being `void(%Foo)*`, but we cannot
+    /// use `%Foo` to compute properties (e.g. size and alignment) of `Foo`,
+    /// until `%Foo` is completed by having all of its field types inserted,
+    /// so `ty` holds the "sizing type" of `Foo`, which replaces all pointers
+    /// with opaque ones, resulting in `{i8*}` for `Foo`.
+    /// ABI-specific logic can then look at the size, alignment and fields of
+    /// `{i8*}` in order to determine how the argument will be passed.
+    /// Only later will `original_ty` aka `%Foo` be used in the LLVM function
+    /// pointer type, without ever having introspected it.
```
Comment (nikomatsakis) [inline src/librustc_trans/trans/abi.rs]: beautiful

**Comment by nikomatsakis** [comment]
@eddyb r=me on the fixups

**Comment by eddyb** [comment]
Re-did the timings, and they look pretty similar.
Old trans:
```
time: 0.248; rss: 79MB  parsing
time: 0.090; rss: 79MB  configuration 1
time: 0.000; rss: 79MB  recursion limit
time: 0.007; rss: 79MB  gated macro checking
time: 0.000; rss: 79MB  crate injection
time: 0.016; rss: 84MB  macro loading
time: 0.000; rss: 84MB  plugin loading
time: 0.000; rss: 84MB  plugin registration
time: 1.154; rss: 121MB expansion
time: 0.041; rss: 121MB complete gated feature checking 1
time: 0.239; rss: 121MB configuration 2
time: 0.000; rss: 121MB gated configuration checking
time: 0.114; rss: 121MB maybe building test harness
time: 0.107; rss: 122MB prelude injection
time: 0.019; rss: 122MB checking that all macro invocations are gone
time: 0.000; rss: 122MB checking for inline asm in case the target doesn't support it
time: 0.041; rss: 122MB complete gated feature checking 2
time: 0.019; rss: 122MB const fn bodies and arguments
time: 0.103; rss: 122MB assigning node ids
time: 0.171; rss: 190MB lowering ast -> hir
time: 0.056; rss: 200MB indexing hir
time: 0.019; rss: 200MB attribute checking
time: 0.069; rss: 200MB early lint checks
time: 0.015; rss: 233MB external crate/lib resolution
time: 0.016; rss: 233MB language item collection
time: 0.284; rss: 286MB resolution
time: 0.025; rss: 286MB lifetime resolution
time: 0.000; rss: 286MB looking for entry point
time: 0.000; rss: 286MB looking for plugin registrar
time: 0.093; rss: 304MB region resolution
time: 0.016; rss: 304MB loop checking
time: 0.016; rss: 304MB static item recursion checking
time: 0.105; rss: 312MB type collecting
time: 0.003; rss: 312MB variance inference
time: 3.870; rss: 332MB coherence checking
time: 0.577; rss: 337MB wf checking
time: 0.365; rss: 343MB item-types checking
time: 11.408; rss: 469MB        item-bodies checking
time: 0.000; rss: 469MB drop-impl checking
time: 0.687; rss: 470MB const checking
time: 0.187; rss: 470MB privacy checking
time: 0.023; rss: 470MB stability index
time: 0.053; rss: 470MB intrinsic checking
time: 0.034; rss: 470MB effect checking
time: 0.173; rss: 470MB match checking
time: 0.099; rss: 471MB liveness checking
time: 0.743; rss: 471MB rvalue checking
time: 0.822; rss: 541MB MIR dump
time: 0.431; rss: 548MB MIR passes
time: 1.373; rss: 549MB borrow checking
time: 0.024; rss: 549MB reachability checking
time: 0.094; rss: 549MB death checking
time: 0.091; rss: 549MB stability checking
time: 0.000; rss: 549MB unused lib feature checking
time: 0.547; rss: 549MB lint checking
time: 0.003; rss: 549MB resolving dependency formats
  time: 3.041; rss: 578MB       translation item collection
  time: 1.211; rss: 1188MB      write metadata
time: 17.216; rss: 1188MB       translation
  time: 5.303; rss: 906MB       llvm function passes [0]
  time: 101.310; rss: 874MB     llvm module passes [0]
  time: 25.368; rss: 891MB      codegen passes [0]
  time: 0.004; rss: 891MB       codegen passes [0]
time: 133.928; rss: 887MB       LLVM passes
  time: 0.345; rss: 887MB       running linker
time: 1.083; rss: 887MB linking
```
MIR trans (`-Z orbit):
```
time: 0.248; rss: 79MB  parsing
time: 0.089; rss: 79MB  configuration 1
time: 0.000; rss: 79MB  recursion limit
time: 0.007; rss: 79MB  gated macro checking
time: 0.000; rss: 79MB  crate injection
time: 0.016; rss: 84MB  macro loading
time: 0.000; rss: 84MB  plugin loading
time: 0.000; rss: 84MB  plugin registration
time: 1.142; rss: 121MB expansion
time: 0.040; rss: 121MB complete gated feature checking 1
time: 0.235; rss: 121MB configuration 2
time: 0.000; rss: 121MB gated configuration checking
time: 0.113; rss: 121MB maybe building test harness
time: 0.105; rss: 122MB prelude injection
time: 0.018; rss: 122MB checking that all macro invocations are gone
time: 0.000; rss: 122MB checking for inline asm in case the target doesn't support it
time: 0.041; rss: 122MB complete gated feature checking 2
time: 0.019; rss: 122MB const fn bodies and arguments
time: 0.101; rss: 122MB assigning node ids
time: 0.169; rss: 191MB lowering ast -> hir
time: 0.055; rss: 200MB indexing hir
time: 0.019; rss: 200MB attribute checking
time: 0.068; rss: 200MB early lint checks
time: 0.015; rss: 233MB external crate/lib resolution
time: 0.016; rss: 233MB language item collection
time: 0.284; rss: 286MB resolution
time: 0.025; rss: 286MB lifetime resolution
time: 0.000; rss: 286MB looking for entry point
time: 0.000; rss: 286MB looking for plugin registrar
time: 0.091; rss: 305MB region resolution
time: 0.016; rss: 305MB loop checking
time: 0.016; rss: 305MB static item recursion checking
time: 0.105; rss: 313MB type collecting
time: 0.003; rss: 313MB variance inference
time: 3.853; rss: 332MB coherence checking
time: 0.569; rss: 337MB wf checking
time: 0.352; rss: 343MB item-types checking
time: 11.255; rss: 469MB        item-bodies checking
time: 0.000; rss: 469MB drop-impl checking
time: 0.687; rss: 471MB const checking
time: 0.185; rss: 471MB privacy checking
time: 0.022; rss: 471MB stability index
time: 0.053; rss: 471MB intrinsic checking
time: 0.033; rss: 471MB effect checking
time: 0.175; rss: 471MB match checking
time: 0.099; rss: 471MB liveness checking
time: 0.740; rss: 471MB rvalue checking
time: 0.835; rss: 542MB MIR dump
time: 0.431; rss: 548MB MIR passes
time: 1.381; rss: 549MB borrow checking
time: 0.024; rss: 549MB reachability checking
time: 0.093; rss: 549MB death checking
time: 0.090; rss: 549MB stability checking
time: 0.000; rss: 549MB unused lib feature checking
time: 0.554; rss: 549MB lint checking
time: 0.003; rss: 549MB resolving dependency formats
  time: 3.044; rss: 578MB       translation item collection
  time: 1.176; rss: 1350MB      write metadata
time: 16.323; rss: 1350MB       translation
  time: 6.786; rss: 1069MB      llvm function passes [0]
  time: 127.018; rss: 1070MB    llvm module passes [0]
  time: 32.242; rss: 1038MB     codegen passes [0]
  time: 0.004; rss: 1007MB      codegen passes [0]
time: 168.301; rss: 1000MB      LLVM passes
  time: 0.352; rss: 1000MB      running linker
time: 1.179; rss: 1001MB        linking
```

@alexcrichton I did check and @dotdash's fixes are in the LLVM build I have, but maybe they're not enough?
The object file size for the MIR build is better, but still noticeably larger:
```
rw-r--r-- 0/0 9153712 Jan  1 00:00 1970 rustc-9026086f.0.o
```

**Comment by eddyb** [comment]
@bors r=nikomatsakis

cc @dotdash I had to re-introduce the newtype-is-immediate-logic, but this time based on the known LLVM type, not some arbitrary Rust type traversal.
Without it, `{i64}` on 32-bit doesn't become an immediate and a very specific test breaks, for atomic and volatile intrinsics, which uses a newtype of `i64` (`isize` would be fine).

**Comment by jonas-schievink** [comment]
```
time: 133.928; rss: 887MB       LLVM passes
```
```
time: 168.301; rss: 1000MB      LLVM passes
```

Am I missing something or does this seem kind of excessive? That's 25% longer, after all...

**Comment by eddyb** [comment]
@jonas-schievink It's doing strictly more work, AFAICT, because we emit more IR (not sure exactly why yet).
As MIR trans is not turned on by default, we can continue to iron out these issues without affecting unsuspecting users.

**Comment by eddyb** [comment]
@bors r=nikomatsakis force

Can we retire that test? For real now, it's been years, why do we still have to deal with it? Especially since it runs after *pretty* much everything else.

**Comment by eddyb** [comment]
@alexcrichton AFAICT the previous failure had to do with volatile operations, but it seemed to segfault the compiler - does that bot not have LLVM assertions turned on?

**Comment by dotdash** [comment]
Actual failure is http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/3455

**Comment by nikomatsakis** [comment]
@eddyb 

> Can we retire that test? For real now, it's been years, why do we still have to deal with it? Especially since it runs after pretty much everything else.

I would not object. Honestly I have no idea what real value it brings. Maybe we can just change it to pretty print something less... intrusive.

**Comment by jonas-schievink** [comment]
:shipit:

-----

> It's doing strictly more work, AFAICT, because we emit more IR (not sure exactly why yet).

Pretty impressive that translation still ends up being faster!

**Pull request merged**